app/abci/utils.go

```diff
@@ -0,0 +1,232 @@
+			}
+			if vote.BlockIdFlag != cmtproto.BlockIDFlagCommit && len(vote.VoteExtension) != 0 {
+				return fmt.Errorf("non-commit vote extension present; validator addr %s",
+					vote.Validator.String(),
+				)
+			}
+			if vote.BlockIdFlag != cmtproto.BlockIDFlagCommit && len(vote.ExtensionSignature) != 0 {
+				return fmt.Errorf("non-commit vote extension signature present; validator addr %s",
+					vote.Validator.String(),
+				)
+			}
+		} else { // vote extensions disabled
+			if len(vote.VoteExtension) != 0 {
+				return fmt.Errorf("vote extension present but extensions disabled; validator addr %s",
+					vote.Validator.String(),
+				)
+			}
+			if len(vote.ExtensionSignature) != 0 {
+				return fmt.Errorf("vote extension signature present but extensions disabled; validator addr %s",
+					vote.Validator.String(),
+				)
+			}
+
+			continue
+		}
+
+		// Only check + include power if the vote is a commit vote. There must be super-majority, otherwise the
+		// previous block (the block vote is for) could not have been committed.
+		if vote.BlockIdFlag != cmtproto.BlockIDFlagCommit {
+			continue
+		}
+
+		// If the validator does not have a valid public key, we skip the signature verification logic but still include
+		// the validator's voting power in the total voting power. The app may have pruned the validator's public key
+		// from the store, but comet considered the validator as active and included them in the commit since there
+		// is a 1 block delay between the validator set update on the app and comet.
+		sumVP += vote.Validator.Power
+		valConsAddr := sdk.ConsAddress(vote.Validator.Address)
+		pubKeyProto, err := valStore.GetPubKeyByConsAddr(ctx, valConsAddr)
+		if err != nil {
+			continue
+		}
+
+		cmtPubKey, err := cryptoenc.PubKeyFromProto(pubKeyProto)
+		if err != nil {
+			return fmt.Errorf("failed to convert validator %X public key: %w", valConsAddr, err)
+		}
+
+		cve := cmtproto.CanonicalVoteExtension{
+			Extension: vote.VoteExtension,
+			Height:    currentHeight - 1, // the vote extension was signed in the previous height
+			Round:     int64(extCommit.Round),
+			ChainId:   chainID,
+		}
+
+		extSignBytes, err := marshalDelimitedFn(&cve)
+		if err != nil {
+			return fmt.Errorf("failed to encode CanonicalVoteExtension: %w", err)
+		}
+
+		if !cmtPubKey.VerifySignature(extSignBytes, vote.ExtensionSignature) {
+			return fmt.Errorf("failed to verify validator %X vote extension signature", valConsAddr)
+		}
+	}
+
+	// This check is probably unnecessary, but better safe than sorry.
+	if totalVP <= 0 {
+		return fmt.Errorf("total voting power must be positive, got: %d", totalVP)
+	}
+
+	if extensionsEnabled {
+		// If the sum of the voting power has not reached (2/3 + 1) we need to error.
+		if requiredVP := ((totalVP * 2) / 3) + 1; sumVP < requiredVP {
+			return fmt.Errorf(
+				"insufficient cumulative voting power received to verify vote extensions; got: %d, expected: >=%d",
+				sumVP, requiredVP,
+			)
+		}
+	}
```

Comment:
Why do we check this only when the extension is enabled?

Reply:
No idea, I took these functions from skip. I don't think it matters much since this function will only be called when they're active so might as well remove it :)